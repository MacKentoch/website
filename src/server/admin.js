import express from 'express'
import objection from 'objection'
import ObjectionRest from 'objection-rest'
import bodyParser from 'body-parser'
import Trainer from 'server/models/Trainer'
import Training from 'server/models/Training'
import Path from 'server/models/Path'
import Course from 'server/models/Course'
import * as database from 'server/services/database'
import path from 'path'

const app = express()

app.use(bodyParser.json())
database.connect()

ObjectionRest(objection)
  .routePrefix('/api')
  .pluralizer(x => x)
  .addModel(Trainer)
  .addModel(Path)
  .addModel(Training)
  .addModel(Course)
  .generate(app)

app.use(express.static(path.join(__dirname, '../../public-admin/dist')))
app.listen(3000)
