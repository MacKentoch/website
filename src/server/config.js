import path from 'path'
import convict from 'convict'

const config = convict({
  env: {
    doc: 'The application environment',
    format: [
      'production',
      'development',
      'test',
      'browser.development',
    ],
    default: 'development',
    env: 'NODE_ENV',
  },
  server: {
    port: {
      doc: 'The server port number',
      format: 'port',
      default: 8000,
      env: 'PORT',
    },
    logFormat: {
      doc: 'The morgan log format to use',
      format: [
        'dev',
        'combined',
        'common',
        'short',
        'tiny',
        '',
      ],
      default: 'dev',
    },
  },
})

const env = config.get('env')
config.loadFile(path.join(__dirname, `../../config/${env}.json`))
config.validate()

export default config
