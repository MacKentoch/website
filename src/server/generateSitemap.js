import sm from 'sitemap'
import Training from 'server/models/Training'
import TrainingSession from 'server/models/TrainingSession'
import { trainingRoute } from 'modules/routePaths'

async function generateSitemap() {
  const trainings = await Training.query().orderByRaw(
    'updated_at desc, id desc',
  )
  const sessions = await TrainingSession.query()
    .eager('[training, location]')
    .whereRaw("start_date > now() + interval '14 day'")
    .orderBy('start_date', 'desc')
  const sessionLinks = await Promise.all(
    sessions.map(session => session.link()),
  )

  const sitemap = sm.createSitemap({
    hostname: 'https://www.smooth-code.com',
    cacheTime: 600000,
    urls: [
      { url: '/', changefreq: 'weekly', priority: 1 },
      { url: '/trainings', changefreq: 'weekly', priority: 0.8 },
      { url: '/story', priority: 0.6 },
      { url: '/contact', priority: 0.6 },
      ...trainings.map(training => ({
        url: trainingRoute(training.slug),
        lastmodISO: training.updated_at.toISOString(),
        changefreq: 'weekly',
        priority: 0.9,
      })),
      ...sessions.map((session, index) => ({
        url: sessionLinks[index],
        lastmodISO: session.updated_at.toISOString(),
        changefreq: 'weekly',
        priority: 0.7,
      })),
    ],
  })

  return new Promise((resolve, reject) => {
    sitemap.toXML((err, xml) => {
      if (err) {
        reject(err)
        return
      }

      resolve(xml.toString())
    })
  })
}

export default generateSitemap
