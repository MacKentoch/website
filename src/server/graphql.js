import Path from 'server/models/Path'
import Training from 'server/models/Training'
import Trainer from 'server/models/Trainer'
import TrainingSession from 'server/models/TrainingSession'
import { GraphQLDate, GraphQLDateTime } from 'graphql-iso-date'
import { makeExecutableSchema } from 'graphql-tools'
import graphqlFields from 'graphql-fields'
import { graphql } from 'graphql'

const resolvers = {
  Date: GraphQLDate,
  DateTime: GraphQLDateTime,
}

export const schema = makeExecutableSchema({
  typeDefs: `
    scalar Date
    scalar DateTime

    type Path {
      id: ID!
      title: String
      color: String
      icon: String
      trainings: [Training]
    }

    type Course {
      id: ID!
      title: String
      outline: String
      path: Path
    }

    type Location {
      id: ID!
      name: String
      address: String
      city: String
      zipcode: String
      country: String
    }

    type Training {
      slug: ID!
      updatedAt: DateTime
      title: String
      longTitle: String
      abstract: String
      socialTitle: String
      socialAbstract: String
      socialPicture: String
      icon: String
      link: String
      duration: Int
      coursePrice: Int
      dayPrice: Int
      intraPrice: Int
      interPrice: Int
      path: Path
      description: String
      objectives: String
      prerequisites: String
      courses: [Course]
      trainers: [Trainer]
      sessions: [Session]
    }

    type Trainer {
      slug: ID!
      fullName: String
      description: String
      picture: String
      link: String
      trainings: [Training]
    }

    type Session {
      id: ID!
      updatedAt: DateTime
      title: String
      abstract: String
      humanizedPeriod: String
      validFrom: Date
      startDate: Date
      endDate: Date
      location: Location
      link: String
      training: Training
    }

    type Query {
      paths: [Path]
      training(slug: ID!): Training
      trainingSession(id: ID!): Session
      trainer(slug: ID!): Trainer

      trainingSessions: [Session]
      trainings: [Training]
    }
  `,
  resolvers,
})

const normalizeQuery = pathOrQuery =>
  typeof pathOrQuery === 'string'
    ? {
        path: pathOrQuery,
        modifiers: {},
      }
    : pathOrQuery

class Eager {
  paths = []
  modifiers = {}

  add(pathOrQuery) {
    const query = normalizeQuery(pathOrQuery)
    this.paths = [...this.paths, query.path]
    this.modifiers = { ...this.modifiers, ...query.modifiers }
  }

  toQuery() {
    if (this.paths.length === 0) return null
    return {
      path: `[${this.paths.join(',')}]`,
      modifiers: this.modifiers,
    }
  }
}

const joinQuery = (source, join) => {
  const sourceQuery = normalizeQuery(source)
  const joinQuery = normalizeQuery(join)

  return {
    path: `${sourceQuery.path}${joinQuery ? `.${joinQuery.path}` : ''}`,
    modifiers: {
      ...sourceQuery.modifiers,
      ...(joinQuery ? joinQuery.modifiers : {}),
    },
  }
}

const eagerResolver = {
  paths(fields) {
    const eager = new Eager()
    if (fields.trainings)
      eager.add(
        joinQuery(
          {
            path: 'trainings',
            modifiers: {
              orderByRank(builder) {
                return builder.orderBy('trainings.rank', 'asc')
              },
            },
          },
          eagerResolver.trainings(fields.trainings),
        ),
      )

    return eager.toQuery()
  },
  trainings(fields) {
    const eager = new Eager()
    if (fields.path) eager.add('path')
    if (fields.trainers) eager.add('trainers')
    if (
      fields.courses ||
      fields.duration ||
      fields.intraPrice ||
      fields.interPrice
    )
      eager.add(
        joinQuery('courses', eagerResolver.courses(fields.courses || {})),
      )
    if (fields.sessions)
      eager.add(
        joinQuery(
          {
            path: 'sessions(liveSessions)',
            modifiers: {
              liveSessions(builder) {
                return builder
                  .whereRaw(
                    "training_sessions.start_date > now() + interval '1 day'",
                  )
                  .orderBy('training_sessions.start_date', 'asc')
                  .limit(3)
              },
            },
          },
          eagerResolver.sessions(fields.sessions),
        ),
      )
    return eager.toQuery()
  },
  courses(fields) {
    const eager = new Eager()
    if (fields.path) eager.add('path')
    return eager.toQuery()
  },
  sessions(fields) {
    const eager = new Eager()
    if (fields.location || fields.link) eager.add('location')
    if (fields.training || fields.link || fields.title || fields.abstract) {
      eager.add(
        joinQuery('training', eagerResolver.trainings(fields.training || {})),
      )
    }
    return eager.toQuery()
  },
  trainers(fields) {
    const eager = new Eager()
    if (fields.trainings)
      eager.add(
        joinQuery('trainings', eagerResolver.trainings(fields.trainings)),
      )
    return eager.toQuery()
  },
}

const enhanceQuery = (query, eagerQuery) => {
  if (eagerQuery) return query.eager(eagerQuery.path, eagerQuery.modifiers)
  return query
}

export const rootValue = {
  async paths(args, obj, context) {
    return enhanceQuery(
      Path.query().orderBy('paths.rank', 'asc'),
      eagerResolver.paths(graphqlFields(context)),
    )
  },
  async training({ slug }, obj, context) {
    return enhanceQuery(
      Training.query().where({ 'trainings.slug': slug }).first(),
      eagerResolver.trainings(graphqlFields(context)),
    )
  },
  async trainingSession({ id }, obj, context) {
    return enhanceQuery(
      TrainingSession.query().where({ 'training_sessions.id': id }).first(),
      eagerResolver.sessions(graphqlFields(context)),
    )
  },
  async trainer({ slug }, obj, context) {
    return enhanceQuery(
      Trainer.query().where({ 'trainers.slug': slug }).first(),
      eagerResolver.trainers(graphqlFields(context)),
    )
  },

  // Sitemap
  async trainingSessions(args, obj, context) {
    return enhanceQuery(
      TrainingSession.query()
        .whereRaw("training_sessions.start_date > now() + interval '14 day'")
        .orderBy('training_sessions.start_date', 'desc'),
      eagerResolver.sessions(graphqlFields(context)),
    )
  },
  async trainings(args, obj, context) {
    return enhanceQuery(
      Training.query().orderByRaw(
        'trainings.updated_at desc, trainings.id desc',
      ),
      eagerResolver.trainings(graphqlFields(context)),
    )
  },
}

export const gql = lines => graphql(schema, lines.join(''), rootValue)
