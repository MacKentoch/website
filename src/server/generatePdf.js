import tmp from 'tmp-promise'
import { createReadStream } from 'fs'
import { exec } from 'mz/child_process'
import config from 'server/config'

async function generatePdf(url) {
  const { path: tmpFile } = await tmp.file()
  const chromeBin = config.get('chrome.bin')
  await exec(
    `"${chromeBin}" --headless --disable-gpu --no-sandbox --print-to-pdf=${tmpFile} ${url}`,
  )
  return createReadStream(tmpFile)
}

export default generatePdf
