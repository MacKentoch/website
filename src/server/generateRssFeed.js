import RSS from 'rss'
import { gql } from 'server/graphql'
import { articlesRoute } from 'modules/routePaths'
import { completeUrl } from 'modules/urlUtil'
import { absClUrl } from 'modules/cloudinary'

async function generateRss() {
  const { data, errors } = await gql`
    {
      articles {
        title
        link
        published_at
        author {
          name
        }
        tags {
          name
        }
        feature_image {
          url
        }
      }
    }
  `

  if (errors) {
    console.error(errors) // eslint-disable-line no-console
    throw new Error('Error during rss generation')
  }

  const feed = new RSS({
    title: 'Smooth Code Actualité',
    description: 'Actualité JavaScript, React et Node.js.',
    generator: 'Smooth Code',
    feed_url: completeUrl('/feed.xml'),
    site_url: completeUrl(articlesRoute()),
    image_url: absClUrl('ujjxt9mthnloobzvxfpa', 'q_auto'),
    managingEditor: 'contact@smooth-code.com',
    language: 'fr',
  })

  data.articles.forEach(article =>
    feed.item({
      title: article.title,
      description: article.custom_excerpt,
      url: completeUrl(article.link),
      categories: article.tags.map(tag => tag.name),
      author: article.author.name,
      date: article.published_at,
    }),
  )

  return feed.xml()
}

export default generateRss
