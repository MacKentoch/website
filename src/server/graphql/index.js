import { buildSchema } from 'graphql'

export const schema = buildSchema(`
  type Training {
    name: String
  }

  type Query {
    training(id: ID!): Training
  }
`)

export const rootValue = {
  training() {
    return { name: 'foo' }
  },
}
