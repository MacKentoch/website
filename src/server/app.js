import path from 'path'
import Koa from 'koa'
import morgan from 'koa-morgan'
import favicon from 'koa-favicon'
import serve from 'koa-static'
import conditional from 'koa-conditional-get'
import etag from 'koa-etag'
import compress from 'koa-compress'
import bodyParser from 'koa-bodyparser'
import error from 'koa-error'
import Router from 'koa-router'
import mount from 'koa-mount'
import Training from 'server/models/Training'
import { graphqlKoa, graphiqlKoa } from 'graphql-server-koa'
import config from 'server/config'
import ssr from 'server/middlewares/ssr'
import redirect from 'server/middlewares/redirect'
import generateSitemap from 'server/generateSitemap'
import generatePdf from 'server/generatePdf'
import sendEmail from 'server/email/sendEmail'
import { schema, rootValue } from 'server/graphql'
import { trainingPdfRoute, trainingPrintRoute } from 'modules/routePaths'

const app = new Koa()
const router = new Router()

app.use(
  redirect([
    {
      match: /^\/trainings\/formation-javascript-es2017/,
      redirect: '/formations/javascript-es2017',
    },
    {
      match: /^\/trainings\/formation-nodejs/,
      redirect: '/formations/nodejs',
    },
    {
      match: /^\/trainings\/formation-react/,
      redirect: '/formations/react',
    },
    {
      match: /^\/trainings\/formation-rxjs/,
      redirect: '/formations/rxjs',
    },
    {
      match: /^\/trainings\/formation-graphql/,
      redirect: '/formations/graphql',
    },
    {
      match: /^\/trainings\/formation-jest/,
      redirect: '/formations/react',
    },
    {
      match: /^\/trainings\/formation-initiation-react/,
      redirect: '/formations/react',
    },
    {
      match: /^\/trainings$/,
      redirect: '/formations',
    },
    {
      match: /^\/story/,
      redirect: '/notre-histoire',
    },
    {
      match: /^\/trainers\/greg-berge/,
      redirect: '/formateurs/greg-berge',
    },
    {
      match: /^\/trainers\/adrien-joly/,
      redirect: '/formateurs/adrien-joly',
    },
    {
      match: /^\/trainers\/thomas-jeanneau/,
      redirect: '/formateurs/thomas-jeanneau',
    },
    {
      match: /^\/creer-app-mac-avec-script-shell/,
      redirect: '/articles/creer-app-mac-avec-script-shell',
    },
    {
      match: /^\/developpez-plus-vite-avec-prettier/,
      redirect: '/articles/developpez-plus-vite-avec-prettier',
    },
    {
      match: /^\/pourquoi-react-est-il-si-populaire/,
      redirect: '/articles/pourquoi-react-est-il-si-populaire',
    },
  ]),
)

router.get('/sitemap.xml', async ctx => {
  ctx.response.type = 'xml'
  ctx.response.body = await generateSitemap()
})

router.post('/api/contact', async ctx => {
  const {
    name,
    company,
    email,
    phone,
    message,
    subject = `${ctx.request.body.name} nous a contactés sur smooth-code.com`,
  } = ctx.request.body
  await sendEmail({
    from: email,
    to: 'contact@smooth-code.com',
    subject,
    textContent: `
${subject}

-----

Nom: ${name}

Société: ${company}

Email: ${email}

Téléphone: ${phone}

-----

${message}
`,
  })
  ctx.body = { error: false }
})

router.get(trainingPdfRoute(':slug'), async ctx => {
  const training = await Training.query()
    .where({ slug: ctx.params.slug })
    .first()

  if (!training) {
    const error = new Error('Training not found')
    error.status = 404
    throw error
  }

  ctx.type = 'pdf'
  ctx.attachment(`${training.slug}-smooth-code.pdf`)
  ctx.body = await generatePdf(
    `${config.get('server.externalUrl')}${trainingPrintRoute(training.slug)}`,
  )
})

const PUBLIC = path.join(__dirname, '../../public')

app.use(error())
app.use(bodyParser())
app.use(serve(PUBLIC, { immutable: true, maxage: 31536000000 }))
app.use(compress({ filter: contentType => /text/i.test(contentType) }))
app.use(conditional())
app.use(etag())

if (config.get('env') !== 'test') {
  app.use(morgan(config.get('server.logFormat')))
}

app.use(favicon(path.join(PUBLIC, 'favicon.ico')))
app.use(router.routes())
app.use(
  mount(
    '/graphql',
    graphqlKoa({
      schema,
      rootValue,
    }),
  ),
)
app.use(mount('/graphiql', graphiqlKoa({ endpointURL: '/graphql' })))
app.use(ssr())

export default app
