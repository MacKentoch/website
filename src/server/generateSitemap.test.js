import { useDatabase, factory } from 'server/test'
import generateSitemap from './generateSitemap'

describe('#generateSitemap', () => {
  useDatabase()

  beforeEach(async () => {
    const paths = await factory.createMany('path', 5, [{ title: 'JavaScript' }])
    await factory.createMany('training', 5, {
      slug: factory.seq(),
      updated_at: '2017-01-01T00:00:00.000Z',
      path_id: paths[0].id,
    })
  })

  it('should generate sitemap', async () => {
    const sitemap = await generateSitemap()
    expect(sitemap).toMatchSnapshot()
  })
})
