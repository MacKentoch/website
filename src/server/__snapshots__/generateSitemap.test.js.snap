// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#generateSitemap should generate sitemap 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\">
<url> <loc>https://www.smooth-code.com/</loc> <changefreq>weekly</changefreq> <priority>1.0</priority> </url>
<url> <loc>https://www.smooth-code.com/formations</loc> <changefreq>weekly</changefreq> <priority>0.8</priority> </url>
<url> <loc>https://www.smooth-code.com/notre-histoire</loc> <priority>0.6</priority> </url>
<url> <loc>https://www.smooth-code.com/contact</loc> <priority>0.6</priority> </url>
<url> <loc>https://www.smooth-code.com/formations/5</loc> <lastmod>2017-01-01T00:00:00.000Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
<url> <loc>https://www.smooth-code.com/formations/4</loc> <lastmod>2017-01-01T00:00:00.000Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
<url> <loc>https://www.smooth-code.com/formations/3</loc> <lastmod>2017-01-01T00:00:00.000Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
<url> <loc>https://www.smooth-code.com/formations/1</loc> <lastmod>2017-01-01T00:00:00.000Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
<url> <loc>https://www.smooth-code.com/formations/2</loc> <lastmod>2017-01-01T00:00:00.000Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
</urlset>"
`;
